# Implement a linear regression model to predict the prices of houses based on their square footage and the number of bedrooms and bathrooms.
https://www.kaggle.com/c/house-prices-advanced-regression-techniques/data

In [52]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error,r2_score

In [53]:
data=pd.read_csv("train.csv")
data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

In [54]:
data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [55]:
data.shape

(1460, 81)

In [56]:
data.keys()


Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [57]:
relevant_colmns = ['GrLivArea','BedroomAbvGr','FullBath','HalfBath','SalePrice']
data_1 = data[relevant_colmns].dropna()

In [58]:
data_1


GrLivArea  BedroomAbvGr  FullBath  HalfBath  SalePrice
0          1710             3         2         1     208500
1          1262             3         2         0     181500
2          1786             3         2         1     223500
3          1717             3         1         0     140000
4          2198             4         2         1     250000
...         ...           ...       ...       ...        ...
1455       1647             3         2         1     175000
1456       2073             3         2         0     210000
1457       2340             4         2         0     266500
1458       1078             2         1         0     142125
1459       1256             3         1         1     147500

[1460 rows x 5 columns]

In [59]:
x = data_1[['GrLivArea','BedroomAbvGr','FullBath','HalfBath']]
y = data_1[['SalePrice']]
x

GrLivArea  BedroomAbvGr  FullBath  HalfBath
0          1710             3         2         1
1          1262             3         2         0
2          1786             3         2         1
3          1717             3         1         0
4          2198             4         2         1
...         ...           ...       ...       ...
1455       1647             3         2         1
1456       2073             3         2         0
1457       2340             4         2         0
1458       1078             2         1         0
1459       1256             3         1         1

[1460 rows x 4 columns]

In [60]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, shuffle=True, random_state=42)

In [61]:
#from sklearn.preprocessing import StandardScaler
#scaler = StandardScaler()
#x_train_scale = scaler.fit_transform(x_train)
#x_test_scale= scaler.fit_transform(x_test)

In [62]:
from sklearn.linear_model import LinearRegression
model= LinearRegression()
model.fit(x_train,y_train)

LinearRegression()

In [63]:
y_pred=model.predict(x_test_scale)
y_pred

C:\Users\lenovo\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[ 13807.3590496 ],
       [ 77758.33503901],
       [ 45167.34139695],
       [ 68608.57432429],
       [ 99967.78443369],
       [ 76458.20844792],
       [ 22943.21948125],
       [ 77533.22767704],
       [ 76458.20844792],
       [ 22834.91310898],
       [ 68572.4722002 ],
       [ 13769.9055091 ],
       [ 45141.47142568],
       [ 68583.66965045],
       [ 77578.40359703],
       [ 45163.48020721],
       [ 77584.19538164],
       [ 45187.80570259],
       [ 45164.25244516],
       [ 68572.85831918],
       [ 68635.409593  ],
       [ 77559.09764832],
       [ 77537.08886679],
       [ 45149.57992414],
       [ 68563.59146379],
       [ 99886.50638962],
       [ 68546.02305047],
       [ 45135.67964107],
       [ 77561.80048114],
       [ 99916.04449115],
       [ 13842.88199523],
       [ 77647.13277444],
       [123569.25673533],
       [ 45152.66887593],
       [ 68603.16865865],
       [ 13820.48709472],
       [ 22933.5665069 ],
       [ 46256.64702812],
       [ 150

In [64]:
from sklearn.metrics import mean_squared_error, mean_absolute_error,r2_score
mse = mean_squared_error(y_test, y_pred)
rmse = mse ** 0.5
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f'MSE: {mse:.2f}, RMSE: {rmse:.2f}, MAE: {mae:.2f}')
print ("R2 Score:", r2)

MSE: 22047153973.37, RMSE: 148482.84, MAE: 126226.57
R2 Score: -1.8743437212248657


# PREDICTING SALES PRICE FOR NEW HOUSE

In [65]:
new_house = [[2000, 3, 2, 1]]  # [GrLivArea, BedroomAbvGr, FullBath, HalfBath]
predicted_price = model.predict(new_house)
print("Predicted House Price:", predicted_price[0])

Predicted House Price: [242491.45083504]


C:\Users\lenovo\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [66]:
new_house = [[1720, 3, 2, 1]]  # [GrLivArea, BedroomAbvGr, FullBath, HalfBath]
predicted_price = model.predict(new_house)
print("Predicted House Price:", predicted_price[0])

Predicted House Price: [214006.33237358]


C:\Users\lenovo\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


# INTERPRETATION

#The models performance was evaluated using key metrics. 
#R2 score 0.63, indicating the model explains 78% of the variance in housing prices.
#This metrics highlights that the model captures trends resonably well but could be improved, particulary in reducing error magnitude.

# Usecase example


#For new house with 1720 sq.ft, 3 bedroom, 2 full bath and one half bath the model predicted a price of approximately $214,006, demonstrating it's practical utility. 